bokehjs/src/lib/client/connection.ts

```diff
@@ -108,51 +103,35 @@ export class ClientConnection {
     message.send(this.socket)
   }
 
-  send_with_reply(message: Message): Promise<Message> {
-    const promise = new Promise<Message>((resolve, reject) => {
+  async send_with_reply(message: Message): Promise<Message> {
+    const reply = await new Promise<Message>((resolve, reject) => {
       this._pending_replies[message.msgid()] = [resolve, reject]
       this.send(message)
     })
 
-    return promise.then(
-      (message) => {
-        if (message.msgtype() === "ERROR")
-          throw new Error(`Error reply ${message.content.text}`)
-        else
-          return message
-      },
-      (error) => {
-        throw error
-      },
-    )
+    if (reply.msgtype() === "ERROR")
+      throw new Error(`Error reply ${reply.content.text}`)
+    else
+      return reply
   }
 
-  protected _pull_doc_json(): Promise<DocJson> {
+  protected async _pull_doc_json(): Promise<DocJson> {
     const message = Message.create("PULL-DOC-REQ", {})
-    const promise = this.send_with_reply(message)
-    return promise.then(
-      (reply) => {
-        if (!('doc' in reply.content))
-          throw new Error("No 'doc' field in PULL-DOC-REPLY")
-        return reply.content.doc
-      },
-      (error) => {
-        throw error
-      },
-    )
+    const reply = await this.send_with_reply(message)
+    if (!("doc" in reply.content))
+      throw new Error("No 'doc' field in PULL-DOC-REPLY")
+    return reply.content.doc
   }
 
-  protected _repull_session_doc(): void {
-    if (this.session == null)
-      logger.debug("Pulling session for first time")
-    else
-      logger.debug("Repulling session")
+  protected _repull_session_doc(resolve: SessionResolver, reject: Rejecter): void {
+    logger.debug(this.session ? "Repulling session" : "Pulling session for first time")
     this._pull_doc_json().then(
```

Comment:
I think it would be good to rewrite this code with async/await as well, just don't use await at the call site of `_repull_session_doc()`. This would make any unit testing easier, as we would be able to access the promise.